### PR TITLE
Build Autocorrect Feature Using NLP with Python

### LINKED ISSUE #4941: Build an Autocorrect Feature using NLP with Python
#### Proposed title of article
Build an Autocorrect Feature using NLP with Python
#### Proposed article introduction
One application of Natural Language Processing (NLP) is the Autocorrect function. If one types a word on the keyboard and the word exists in the vocabulary of that smartphone device then it will generalize the word as the correct word. If the word is non existing in the vocabulary, then the autocorrect feature is specially programmed to find the most similar words in the vocabulary as it suggests. 
To understand the working and how to build this feature, this article will equip the reader to understand Natural Language Processing and how it's used to build an autocorrect model with python.
#### Key takeaways 
The reader will be able to:
- Understand Natural Language Processing (NLP) and some of its applications.
- Understand the steps involved to build the autocorrect feature.
- Understand how to build the autocorrect feature using python in Natural Language Processing.

#### Article quality
This will be of high quality different from the others, as it not only shows the NLP applications but also its implementation with Python in building the various applications.

#### References
N/A
### Conclusion
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by dentex22**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This fixes #4941

### CODE REVIEW

**Comment by okeloviolet** [comment]
@hectorkambow I will be reviewing this article

**Comment by ahmadmardeni1** [comment]
Please always use the template provided by us.
I edited your initial comment and added the template. Check what matches your article.

**Comment by dentex22** [comment]
> Please always use the template provided by us. I edited your initial comment and added the template. Check what matches your article.

@ahmadmardeni1 any updates ??

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+author: antony-lia
+date: 2021-11-13T00:00:00-2:30
+topics: [Machine Learning]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/build-autocorrect-feature-using-nlp-with-python/hero.jpg
+    alt: Natural Language processing example image
+---
+One application of Natural Language Processing (NLP) is the Autocorrect function. This feature works on every smartphone keyboard regardless of the brand, it is specially programmed to generalize all the correct words that are in the vocabulary and tries to find the most similar words to those words, not in the vocabulary.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:17]: remove the comma between brand and it. Replace "it is specially programmed to generalize all the correct words that are in the vocabulary and tries to find the most similar words to those words, not in the vocabulary." with " It is specially programmed to generalize all the correct words in the vocabulary and try to find the most similar words to those words, not in the vocabulary.	"

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+<!--more-->
+To understand how it works, will cover Natural Language Processing in this article and how we can use it with Python to build the autocorrect feature showing all the steps involved.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:19]: Change to " To understand how it works, we will cover Natural Language Processing in this article and use it with Python to build the autocorrect feature showing all the steps involved."

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  - [Finding Similar/Corrected Word](#finding-similarcorrected-word)
+- [Conclusion](#conclusion)
+
+### Prerequisites
+To follow along with this tutorial, the reader should:
+- Have a prior understanding of the basics of Natural Language Processing and Machine Learning concepts.
+- Have a basic understanding of Python and the various python libraries used in Natural language processing.
+- Know how to use Pycharm or any other IDE or code editor for running Python.
+
+To download and install the Pycharm IDE on your PC, go to this [site](https://www.jetbrains.com/pycharm/download/#section=windows).
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:40]: Change to " Go to this [site] to download and install the Pycharm IDE on your PC, go to this [site](https://www.jetbrains.com/pycharm/download/#section=windows)."

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+
+### Natural Language Processing(NLP)
+Natural Language Processing (NLP) is a branch of computer science and artificial intelligence that focuses on allowing computers to understand and process human natural language.. NLP is a programming language that allows computers to evaluate and interpret large volumes of natural language data.
+It paves the door for more interactivity and productivity in a variety of fields:
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:45]: Change to " Natural Language Processing (NLP) is a branch of computer science and artificial intelligence that allows computers to understand and process natural human language. NLP is a programming language that enables computers to evaluate and interpret large volumes of natural language data."

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+- Search Autocorrect and Autocomplete.
+- Language Translation and Grammar checkers.
+- Chatbots and Social Media monitoring.
+- Email filtering and Voice Assistants.
+  
+We'll look at how it's employed in autocorrection systems in this tutorial.
+
+### Autocorrect Feature
+The Autocorrect model is completely based on Natural Language Processing (NLP), and it is programmed to correct spellings and errors while inputting text and tries to locate the most comparable related words, as the name suggests. It works by comparing the words in the vocabulary dictionary and the typed words on the keyboard.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:54]: Change to " The Autocorrect model is completely based on Natural Language Processing (NLP). It is programmed to correct spellings and errors while inputting text and locating the most comparable related words, as the name suggests. It compares the words in the vocabulary dictionary and the typed words on the keyboard. "

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+ - *Find Strings that are n Edit distance away from misspelled word* - editing is an operation  performed on the string to change it to another string. The `n` represents the edit distance like 1,2,3...which will keep track of the number of edit operations that need to be done. Hence the `edit distance` is the count of the number of operations performed on a word to edit it.
+The following are examples of edits:
+      `INSERT` - a letter should be added.
+      `DELETE` - remove a letter.
+      `SWAP` - swap to adjacent letters.
+      `REPLACE` - changes one letter to another.
+  
+  >NOTE: We usually take n between 1 to 3 edits for autocorrect systems.
+
+ - *Filtering Suggested Candidates* - Only correctly spelt real words from our created candidate list are considered, so we can compare them to the words in the dictionary and then filter out the ones that don't exist in the dictionary.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:68]: Change "spelt" to "spelled"

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+```bash
+The first ten words in the text are: 
+['a', 'ability', 'able', 'about', 'above', 'accept', 'according', 'account', 'across', 'act']
+There are 1001 unique words in the vocabulary.
+```
+#### Frequency and Probability of Dictionary Words 
+We use the `counter function` to find the frequency of the words by applying the code below.
+```python
+# a get_count function that returns a dictionary of word vs frequency
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:124]: Change to " # a get_count function that returns a dictionary of word versus frequency"

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+def get_probs(word_count_dict):
+    probs = {}
+    m = sum(word_count_dict.values())
+    for key in word_count_dict.keys():
+        probs[key] = word_count_dict[key] / m
+    return probs
+```
+#### Implement 4 edit word functions
+Each of the 4 edit functions is implemented as shown below each performing different tasks as early illustrated. We can also combine the functions to be able to find a list of all the strings that are n edits to perform.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:158]: Change to " Each of the four edit functions is implemented below, each performing different tasks as illustrated earlier. We can also combine the functions with finding a list of all the strings that are n edited to execute."

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    edit_two_set = set()
+    edit_one = edit_one_letter(word, allow_switches=allow_switches)
+    for w in edit_one:
+        if w:
+            edit_two = edit_one_letter(w, allow_switches=allow_switches)
+            edit_two_set.update(edit_two)
+    return edit_two_set
+```
+#### Finding Similar/Corrected Word
+The program prompts the user to enter a word, if the word exists in the dictionary it is assumed to be right else a similar word is produced by the program. This is implemented by this piece of code.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:240]: Change to "The program prompts the user to enter a word. If the word exists in the dictionary, it is assumed to be right, or else the program produces a similar word. This piece of code implements this."

File: content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+['a', 'ability', 'able', 'about', 'above', 'accept', 'according', 'account', 'across', 'act']
+There are 1001 unique words in the vocabulary.
+There are 1001 key values pairs
+Enter a word:daed
+word 0: dead, probability 0.000999
+
+Process finished with exit code 0
+```
+### Conclusion
+As we have seen, Natural Language Processing plays a crucial role in enabling computers to understand and process human natural language. This is as implemented above using the autocorrect system.
```
Comment (okeloviolet) [inline content/articles/build-autocorrect-feauture-using-nlp-with-python/index.md:272]: Change "human natural language" to "natural human language"

**Review by okeloviolet — CHANGES_REQUESTED** [review]
Hello @dentex22 Make the request changes, I will proceed to tech review.

**Comment by dentex22** [comment]
> Hello @dentex22 Make the request changes, I will proceed to tech review.

ook

**Comment by dentex22** [comment]
@okeloviolet done with changes

**Comment by dentex22** [comment]
> @okeloviolet done with changes

label update?? @okeloviolet

**Comment by ahmadmardeni1** [comment]
👋 Good afternoon - and welcome to the EngEd program.
There appear to be many sentences in this article that are taken verbatim from third-party sources. We can't publish any content that is not original or properly cited/referenced. There are many free online tools where you can check this (such as quetext.com, and if you're uncertain about the definitions around plagiarism, this is a [good resource to reference](https://www.scribbr.com/category/plagiarism/).
We're really looking for unique/original content. Take your time and revisit your article. @dentex22 

[New Scan 9_08 AM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7561562/New.Scan.9_08.AM.Copyleaks.Report.pdf)

**Pull request closed without merging**